### DISCUSSION TITLE
How to approach asyn derived atom if the data and params to query it not known upfront.

### DISCUSSION THREAD

**Discussion by alexshelkov** [pmndrs/jotai#2352] [Q&A] [answered]
Lets say I have some atom with current page:

```typescript
const page = atom<string | null>(null);
```

By default I don't know the page, lets say it I fetch it from the server.

Next I have a collection of comments for this page:

```typescript
const comments = atom(
  async (get) => {
    const id = get(page);
    
    if (!id) {
      // the atom should be at loading state
      // something like new Promise() which never resolves
      return // WHAT?
    }
    
    require ["Comment 1", "Comment 2"];
  }
)
```
I am not sure how approach to this problem, if I return an empty array (if there is no id) it means that atom will be resolved. But this is not the thing which I really want.

**Comment by dai-shi** [comment]
You can't do like that.

What you can do is something like this: https://jotai.org/docs/guides/async#async-sometimes
This blog post might also help: https://blog.axlight.com/posts/how-to-use-jotai-and-use-transition-for-mutation/

**Reply by alexshelkov** [reply]
Sorry, but I tried to understand how the articles might help me, but I just can't get it. Could you pls give me some example of how achieve behavior I want?

**Reply by dai-shi** [reply, marked as answer]
okay, forget about the article.

The short answer is that you can't do what you want.
What could be done instead is to store a promise value.
```js
const internalPageAtom = atom(null);
const commentsAtom = atom(new Promise(() => {}));
const pageAtom = atom(
  (get) => get(internalPageAtom),
  (get, set, newPage) => {
    set(internalPageAtom, newPage);
    set(commentsAtom, ['comment1', 'comment2']);
  }
);
```

**Reply by alexshelkov** [reply]
So this becomes very imperative, and if I have few other things which depends on the `page`, it means that I have to update it in the `pageAtom`, which becomes more and more big?

And thanks for help!

**Reply by dai-shi** [reply]
yes, it's only possible to update the infinite promise imperatively.
Other atoms can still be declarative:
```js
const existsPageAtom = atom((get) => !!get(pageAtom));
```
